### ISSUE TITLE
Defining the local coordinate system of elements (beams, shells)

### ISSUE THREAD

**Issue by AndreasWinterstein** [KratosMultiphysics/Kratos#136]
Hi together,

with this post the discussion for the definition of the local coordinate system for elements, which is especially crucial for beam elements, is opened. Feel free to add other members of the team who might be of interested.

Here is our suggestion:

The local x-axis is the vector spanning from the starting point to the end point of the beam. Then the local y-axis is calculated with help of the cross product (gobalZ(0,0,1) X local x-axis). Another crossproduct (local x-axis X local y-axis) results in the local z-axis. All local axis are normalized.
One exception:
In case that the beam axis is parallel to the global Z-Axis:
nX = (0, 0, +- 1); nY = (0, 0 ,1); nZ = (-+ 1, 0, 0)

Looking forward to your comments.

Andreas

**Comment by RiccardoRossi** [comment]
@AndreasWinterstein, i don't really love exceptions :-D

let's assume that v1, v2, v3 are unit vectors

what would feel natural to me is:

**BEAM CASE** (with 2 nodes of coordinates x1, x2)
v3 := (x2 - x1)/norm(x2-x1) - automatically defined by the geometry
v2 := user defined (must be orthogonal to v3)
v1 := v3 x v2 (x indicate cross prod)

beam in the XY plane, with axis parallel to X:
v3 = (1,0,0) v2 = (0,1,0) v1 = (0,0,1)

beam in the YZ plane with axis parallel to Z
v3 = (0,0,1) v2 = (0,1,0) v1 =( 0,0,-1)

@AndreasWinterstein ,where is the "special case" needed? 
Massimo, what is the ordering that Abacus employs? what about OpenSEES?

**SHELL CASE:**
here the normal is well defined. so we need at least 1 tangent vector
let's say that
v3 = normal - automatically defined
v1 = tangent vector
v2 := v3 x v1 (note that ordering is important here)
this way vector 1 and 2 belong to the tangent plane and the third direction is normal to it

**GENERAL 3D case**
user must specify 2 vectors, say v1 and v2
v1 = user specified
v2 = user specified (must be orthogonal to v1)
v3 := v1 x v2

@jginternational what would be the orientation that fits more naturally to GiD? please also ask to Kike

**Comment by loumalouomega** [comment]
@RiccardoRossi I asked to the GiD people and they use the Euler Angles to define the Local Axes.

According to GiD doc:

![](https://i.imgur.com/McXYuvl.png)

I implemented a conversion from orthogonal base to Euler angles the other day, here the  code:

~~~c
    /**
     * This function recovers the Euler angles from an orthogonal base (Local Coordinate System)
     * @param Normal: The normal vector
     * @param TangentXi: The first tangent vector
     * @param TangentEta: The second tangent vector
     * @return angles[0]: (pre) precession rotation
     * @return angles[1]: (nut) nutation rotation
     * @return angles[2]: (rot) intrinsic rotation
     */
    
    array_1d<double, 3> LCS2Euler(
        const array_1d<double, 3> Normal,
        const array_1d<double, 3> TangentXi,
        const array_1d<double, 3> TangentEta
        ) 
    {
        const double TangentEtaxy = std::sqrt(TangentEta[0] * TangentEta[0] + TangentEta[1] * TangentEta[1]);
        
        array_1d<double, 3> angles;
        
        if (TangentEtaxy > std::numeric_limits<double>::epsilon()) 
        {
            angles[0] =  std::atan2(TangentXi[0] * TangentEta[1] - TangentXi[1]*TangentEta[0], Normal[0] * TangentEta[1] - Normal[1] * TangentEta[0]);
            angles[1] =  std::atan2(TangentEtaxy, TangentEta[2]);
            angles[2] = -std::atan2(-TangentEta[0], TangentEta[1]);
        }
        else 
        {
            angles[0] = 0.0;
            angles[1] = (TangentEta[2] > 0.0) ? 0.0 : M_PI;
            angles[2] = -std::atan2(Normal[1], Normal[0]);
        }

        return angles;
    }
~~~

**Comment by RiccardoRossi** [comment]
@loumalouomega i would prefer NOT to use Euler angles, since they are typically plagued by a lot of special cases/different convenctions.

prescribing the vectors directly would be way more robust...

**Comment by msandre** [comment]
I agree with @RiccardoRossi for the beam case. Allowing the user definition of v2 would make defining area moment of inertia simpler.

An exception free, numerically stable way to generate a coordinate system automatically from one vector (beam axis or surface normal) is:

v := starting vector 
v3 := v / norm(v)
t2 := global axis i such that abs(v3_i) <= abs(v3_k) for k=1..dim
v2 := (t2 - v3_i * v3) / norm(t2 - v3_i * v3) // gram-schmidt orthonormalization
v1 := v3 x v2

Maybe this is still useful for axisymmetic beams when the user doesn't want to calculate a tangent. 

The code for v2 looks like:
unsigned int i = 0;
for (unsigned int d = 0; d < TDim; ++d) 
{
  v2[d] = 0.0;
  if (std::abs(v3[d]) < std::abs(v3[i]))
    i = d;
}

v2[i] = 1.0;

for (unsigned int d = 0; d < TDim; ++d)
  v2[d] -= v3[i] * v3[d];

v2 /= norm(v2);

**Comment by josep-m-carbonell** [comment]
GiD has a convention to define local axes in beam elements. This is the one that has been used for beams. You can locate it in the small displacement beam element. Ask Massimo what is the convention used in shells. I think that if there is no major problem to use the current ones we don't have to change them or invent new ones.

**Comment by josep-m-carbonell** [comment]
Usually, the user have an option to use the automatic definition of local axes. This is the convention by default. But there is also tge options of customize it for each element. Then you have to supply the local axes definition to each element, via properties or via specialized constructor. The last option (input mechanism) is not defined yet in Kratos. It is a good issue to discuss about.

**Comment by loumalouomega** [comment]
@josep-m-carbonell Do you mean this?

~~~c
void SmallDisplacementBeamElement3D2N::CalculateTransformationMatrix(Matrix& rRotationMatrix)

{

    KRATOS_TRY
    const unsigned int number_of_nodes = GetGeometry().size();
    const unsigned int dimension       = GetGeometry().WorkingSpaceDimension();
    unsigned int       size            = number_of_nodes * dimension;
    unsigned int       MatSize         = 2 * size;


    Vector GlobalY = ZeroVector(3);
    GlobalY[1]=1.0;

    Vector GlobalZ = ZeroVector(3);
    GlobalZ[2]=1.0;

    Vector DirectionVectorX     = ZeroVector(3);
    Vector ReferenceCoordinates = ZeroVector(size);


    ReferenceCoordinates[0] = GetGeometry()[0].X();
    ReferenceCoordinates[1] = GetGeometry()[0].Y();
    ReferenceCoordinates[2] = GetGeometry()[0].Z();

    int k = number_of_nodes - 1 ;

    ReferenceCoordinates[3] = GetGeometry()[k].X();
    ReferenceCoordinates[4] = GetGeometry()[k].Y();
    ReferenceCoordinates[5] = GetGeometry()[k].Z();

    for( unsigned int i = 0; i < dimension; i++ )
    {
        DirectionVectorX[i]  = (ReferenceCoordinates[i+3] - ReferenceCoordinates[i]);
    }

    // local x-axis (e1_local) is the beam axis  (in GID is e3_local)
    double VectorNorm = MathUtils<double>::Norm(DirectionVectorX);
    if( VectorNorm != 0)
        DirectionVectorX /= VectorNorm;

    // local y-axis (e2_local) (in GID is e1_local)
    Vector DirectionVectorY = ZeroVector(3);

    double tolerance = 1.0/64.0;
    if(fabs(DirectionVectorX[0])< tolerance && fabs(DirectionVectorX[1])< tolerance)
    {
        DirectionVectorY = MathUtils<double>::CrossProduct(GlobalY, DirectionVectorX);
    }
    else
    {
        DirectionVectorY = MathUtils<double>::CrossProduct(GlobalZ, DirectionVectorX);
    }

    VectorNorm = MathUtils<double>::Norm(DirectionVectorY);
    if( VectorNorm != 0)
        DirectionVectorY /= VectorNorm;

    // local z-axis (e3_local) (in GID is e2_local)
    Vector DirectionVectorZ = MathUtils<double>::CrossProduct(DirectionVectorX,DirectionVectorY);

    VectorNorm = MathUtils<double>::Norm(DirectionVectorZ);
    if( VectorNorm != 0 )
        DirectionVectorZ /= VectorNorm;


    //Transformation matrix T = [e1_local, e2_local, e3_local]
    Matrix AuxRotationMatrix;

    if( AuxRotationMatrix.size1() != dimension )
        AuxRotationMatrix.resize(dimension, dimension, false);

    // std::cout<<" Xlocal "<<DirectionVectorX<<std::endl;
    // std::cout<<" Ylocal "<<DirectionVectorY<<std::endl;
    // std::cout<<" Zlocal "<<DirectionVectorZ<<std::endl;

    for (unsigned int i=0; i<dimension; i++)
    {
        AuxRotationMatrix(i,0) = DirectionVectorX[i];  // column distribution
        AuxRotationMatrix(i,1) = DirectionVectorY[i];
        AuxRotationMatrix(i,2) = DirectionVectorZ[i];
    }


    if( rRotationMatrix.size1() != MatSize )
        rRotationMatrix.resize(MatSize, MatSize, false);

    rRotationMatrix = ZeroMatrix(MatSize);

    //Building the rotation matrix for the local element matrix
    for (unsigned int kk=0; kk < MatSize; kk += dimension)
    {
        for (unsigned int i=0; i<dimension; i++)
        {
            for(unsigned int j=0; j<dimension; j++)
            {
                rRotationMatrix(i+kk,j+kk) = AuxRotationMatrix(i,j);
            }
        }
    }


    KRATOS_CATCH( "" )

}
~~~

**Comment by josep-m-carbonell** [comment]
Yes, this is the same operation that GiD perform when the automatic axes(1) is selected. I think there is also another automatic definition for local axes suplied by GiD, ask Quique. Maybe I have a more optimal implementation in my beam elements. I have to check it.  Anyway, the result will be the same.

**Comment by loumalouomega** [comment]
@RiccardoRossi Tou asked about other FEA software criteria:
Abaqus: http://abaqus.software.polimi.it/v6.13/books/usb/default.htm?startat=pt01ch02s02aus15.html
Ansys: http://www.ansys.stuba.sk/html/guide_55/g-mod/gmod3.htm

**Comment by MassimoPetracca** [comment]
Dear all,

Here we have 2 problems:
1) how to get the user-defined local axes
2) how to handle the "default-cases" in which no user-defined local axis is given

**1) how to get the user-defined local axes**
Since each pre-processor may have its own way of representing an orientation (one of the many Euler angles convention, Quaternions, etc..), I would use the most general one: 
x,y and z vectors. Or, to save space, just 2 of them, x and y, and the 3rd can be computed if needed.
Then I would define variables in kratos, so that they can be inputed as properties.

**2) how to handle the "default-cases" in which no user-defined local axis is given**
**Shell case:**
For a shell, the only direction we know for sure is the normal (**z'**). Then we need to compute the in plane directions (**x'** and **y'**).
```cpp
if(std::abs(local_z[2]) > 1.0-tolerance) { // local z' aligned with global Z
    local_x = vector3(1, 0, 0);
}
else {
    local_x = cross_prod( global_Z, local_z ); 
}
local_y = cross_prod( local_z, local_x );
```
The result of this convention is that, at least for simple structures (i.e. simple buildings) with vertical walls, and planar slabs, planar slabs have local axes aligned with the global ones, while vertical walls have local x' axis aligned counter-clock-wise about the global Z, and local y' aligned along the global Z.

**Beam case:**
For a beam, the only direction we know for sure is the beam axis (**x'**). Then we need to compute the orientation of the cross section (**y'**, **z'**) about the beam axis. Typically (Ansys, MidasGEN, Abaqus?) the horizontal axis of a cross-section is the y', while the z' is the vertical one:
![image](https://cloud.githubusercontent.com/assets/26160906/23793969/1e9e2592-058e-11e7-80fd-bd36fb3ae429.png)
(**_Note: this is another convention we need to specify_**)
```cpp
if(std::abs(local_x[2]) > 1.0-tolerance) { // local x' aligned with global Z (vertical beam)
    local_z = vector3(1, 0, 0);
    local_y = cross_prod( local_z, local_x );
}
else {
    local_y = cross_prod( global_Z, local_x );
    local_z = cross_prod( local_x, local_y );
}
```
![image](https://cloud.githubusercontent.com/assets/26160906/23794292/b8fdc07e-058f-11e7-8950-359a9e73e143.png)

**Comment by RiccardoRossi** [comment]
Hello everyone.

i am writing here to try to resume. I see already that we are all used to different convenctions for the beam :-(

let's begin for the case in which WE DO WANT To prescribe explicitly the orientation.

i would propose to create two variables of type array_1d<double,3>
DIRECTION1
DIRECTION2
I think that it is **not** necessary to also have DIRECTION3, since it can always be computed using the first two.

**BEAM CASE**
user should define DIRECTION2 which represents the direction of axis "2" (y in Massimo's picture)
we need to decide if the direction of the beam is 
a) DIRECTION1 is considered as the direction of the beam (as Massimo Proposes)
b) DIRECTION3 is considered to be the direction of the beam (as Andreas suggested)
in either case i think that DIRECTION3 = DIRECTION1 x DIRECTION2
can we make a choice between a) and b)? 
I also would like to say that different books use different convenctions, so that there is not "A GOOD ONE", so let's just try to find an agreement. In particular the book on the basis of which i teach chooses option1, but then inverts the order of the axes so i think it is not suitable here.

what to do in the default case (which is essentially how to choose by default what i am calling DIRECTION2) follows from this decision, so let's first clarify this point

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi I think that in your resume the X direction is not correct, while everybody except you agrees that they should be aligned with beam direction. (You mentioned that this would be the direction3)

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand 

let's say that XYZ are the global directions
then 1 2 3 are the directions corresponding to the local axes v1 v2 v3. I am using numbers instead of symbols not to make confusion.

If you are saying that direction v1 should be the direction along the beam axis i am fine with that. (it is option a) of the two i listed.

then the user should define a vector which describes the orientation of 2, that is, assign a vector
DIRECTION2.

then 
v1 = PositionNode2 - PositionNode1 (normalized) -- THIS IS DIRECTION X in Massimo's figure
v2 = DIRECTION2 (user defined) -- THIS IS DIRECTION Y in Massimo's figure
v3 = v1 x v2 -- THIS IS DIRECTION Z in Massimo's figure

is there agreement on this?

**Comment by josep-m-carbonell** [comment]
What is wrong in using the convention that is already programed and is used right now? . If you change it you will break the current implementation.

**Comment by RiccardoRossi** [comment]
could you explain here how is the current convention?

**Comment by loumalouomega** [comment]
@RiccardoRossi It is the one I posted, which comes from the BeamElement3D2N

**Comment by RiccardoRossi** [comment]
@loumalouomega i personally think that Euler angles are no go... my opinion though.

in any case could u explain the convention without code? just in symbols?
how would u prescribe that DIRECTION2 is (-1/sqrt(2), -1/sqrt(2),0) (with the beam axis is oriented in the Z direction)

a big advantage of having a direction vector, is that it can be visualized and assigned as one likes. 

@josep-m-carbonell is it just one element affected or there are more?

**Comment by loumalouomega** [comment]
@RiccardoRossi , I meant this:

~~~c
void SmallDisplacementBeamElement3D2N::CalculateTransformationMatrix(Matrix& rRotationMatrix)

{

    KRATOS_TRY
    const unsigned int number_of_nodes = GetGeometry().size();
    const unsigned int dimension       = GetGeometry().WorkingSpaceDimension();
    unsigned int       size            = number_of_nodes * dimension;
    unsigned int       MatSize         = 2 * size;


    Vector GlobalY = ZeroVector(3);
    GlobalY[1]=1.0;

    Vector GlobalZ = ZeroVector(3);
    GlobalZ[2]=1.0;

    Vector DirectionVectorX     = ZeroVector(3);
    Vector ReferenceCoordinates = ZeroVector(size);


    ReferenceCoordinates[0] = GetGeometry()[0].X();
    ReferenceCoordinates[1] = GetGeometry()[0].Y();
    ReferenceCoordinates[2] = GetGeometry()[0].Z();

    int k = number_of_nodes - 1 ;

    ReferenceCoordinates[3] = GetGeometry()[k].X();
    ReferenceCoordinates[4] = GetGeometry()[k].Y();
    ReferenceCoordinates[5] = GetGeometry()[k].Z();

    for( unsigned int i = 0; i < dimension; i++ )
    {
        DirectionVectorX[i]  = (ReferenceCoordinates[i+3] - ReferenceCoordinates[i]);
    }

    // local x-axis (e1_local) is the beam axis  (in GID is e3_local)
    double VectorNorm = MathUtils<double>::Norm(DirectionVectorX);
    if( VectorNorm != 0)
        DirectionVectorX /= VectorNorm;

    // local y-axis (e2_local) (in GID is e1_local)
    Vector DirectionVectorY = ZeroVector(3);

    double tolerance = 1.0/64.0;
    if(fabs(DirectionVectorX[0])< tolerance && fabs(DirectionVectorX[1])< tolerance)
    {
        DirectionVectorY = MathUtils<double>::CrossProduct(GlobalY, DirectionVectorX);
    }
    else
    {
        DirectionVectorY = MathUtils<double>::CrossProduct(GlobalZ, DirectionVectorX);
    }

    VectorNorm = MathUtils<double>::Norm(DirectionVectorY);
    if( VectorNorm != 0)
        DirectionVectorY /= VectorNorm;

    // local z-axis (e3_local) (in GID is e2_local)
    Vector DirectionVectorZ = MathUtils<double>::CrossProduct(DirectionVectorX,DirectionVectorY);

    VectorNorm = MathUtils<double>::Norm(DirectionVectorZ);
    if( VectorNorm != 0 )
        DirectionVectorZ /= VectorNorm;


    //Transformation matrix T = [e1_local, e2_local, e3_local]
    Matrix AuxRotationMatrix;

    if( AuxRotationMatrix.size1() != dimension )
        AuxRotationMatrix.resize(dimension, dimension, false);

    // std::cout<<" Xlocal "<<DirectionVectorX<<std::endl;
    // std::cout<<" Ylocal "<<DirectionVectorY<<std::endl;
    // std::cout<<" Zlocal "<<DirectionVectorZ<<std::endl;

    for (unsigned int i=0; i<dimension; i++)
    {
        AuxRotationMatrix(i,0) = DirectionVectorX[i];  // column distribution
        AuxRotationMatrix(i,1) = DirectionVectorY[i];
        AuxRotationMatrix(i,2) = DirectionVectorZ[i];
    }


    if( rRotationMatrix.size1() != MatSize )
        rRotationMatrix.resize(MatSize, MatSize, false);

    rRotationMatrix = ZeroMatrix(MatSize);

    //Building the rotation matrix for the local element matrix
    for (unsigned int kk=0; kk < MatSize; kk += dimension)
    {
        for (unsigned int i=0; i<dimension; i++)
        {
            for(unsigned int j=0; j<dimension; j++)
            {
                rRotationMatrix(i+kk,j+kk) = AuxRotationMatrix(i,j);
            }
        }
    }


    KRATOS_CATCH( "" )

}
~~~

**Comment by josep-m-carbonell** [comment]
By default in GiD the axis of the beam is e3, and then the orthogonal ones are e1 and e2 obtained by cross products depending on the axis direction. The algorithm is in the code above-mentioned. 

In most beam formulations the axis of the beam is e1,  so in the kratos elements,  the axis is e1,  and the orthogonal are e2 and e3. These (e2, e3)  are obtained following the same algorithm as GiD for the orthogonal axis. So finally the corresondence is Kratos(e1, e2, e3) are GiD(e3,e1,e2).

**Comment by josep-m-carbonell** [comment]
If you look at the code, the most tricky thing is the definition of kratos e2,  because checks the axis vertical or horizontal direction. But if you skip that you get strange disposition of the axes by default. Remember that then this affects to the direction of the beam seccion considered.

**Comment by josep-m-carbonell** [comment]
And yes, there are other elements outside the StructuralMechanicsApplication that uses the these convention and will break completely if it is changed.

**Comment by loumalouomega** [comment]
Which elements?, are they accessible in any public application?

**Comment by josep-m-carbonell** [comment]
They are not public yet. They belong to a specific aplication for a project called StringDynamicsApplication.

**Comment by msandre** [comment]
All the different notations are causing a lot of confusion. Could we try to speak in one notation? I will try to summarize with a single notation v1,v2,v3 originally from @RiccardoRossi and global axes ex,ey,ez.

We have the following cases:

1) Using GiD code of @josep-m-carbonell (numerically stable version of what @AndreasWinterstein proposed):
v3 = DirectionVectorX (Beam Axis)
v2 = DirectionVectorY
v1 = DirectionVectorZ

2) @MassimoPetracca Shell Case
v3 = local_z (Shell Normal)
v2 = local_y
v1 = local_x

3) @MassimoPetracca Beam Case
v3 = local_x (Beam Axis)
v2 = local_y
v1 = local_z

Now if v3 ~ ez we have:

Case 1) v2 = ey x v3 ~ ey x ez ~ ex and v1 = v3 x v2 ~ ez x ex ~ ey
Case 2) v1 = ex and v2 = v3 x v1 ~ ez x ex ~ ey
Case 3) v1 = ex and v2 = v1 x v3 ~ ex x ez ~ -ey

Othewise:

Case 1) v2 = ez x v3 and v1 = v3 x v2
Case 2) v1 = ez x v3 and v2 = v3 x v1
Case 3) v2 = ez x v3 and v1 = v3 x v2

My only objection to 2) & 3) is that the axes are not exactly orthogonal. Further, I would generate the v2 using the alternative orthonormalization instead of the cross product. You will get the same axes but it should be less computational effort.

E.g. for case 2:
v3 ~ ez
t2 := ey
v2 := (t2 - v3_i * v3) / norm(t2 - v3_i * v3) ~ ey
v1 := v2 x v3 ~ ex

**Comment by RiccardoRossi** [comment]
@msandre there is one detail: in order for the basis to be oriented consistently, the following relations should be verified 

    v1 x v2 = v3
    v2 x v3 = v1
    v3 x v1 = v2

if this is not so, the matrix obtained putting in columns (v1,v2,v3) will not be a rotation matrix.

Aside of this in any case for a second about the implementation.
here, as you say, i am calling v1,v2,v3 the unit vectors corresponding to the local axis
since i suppose that e1,e2,e3 are the unit vectors corresponding to the global system.
i also supposed here that the beam axis is oriented as v3, however please do not focus on that and let's agree on the idea first.

imagine that we have somewhere (in common) the following functions, the implementation of which is what i would like to discuss here


    ##user-prescribed case
    void ComputeBeamOrientation( array_1d<double,3>& v1, //output parameter: 
                             const array_1d<double,3>& v2, //input parameter: direction of 2
                             array_1d<double,3>& v3        //input parameter direction of the beam axis
                       )

     void ComputeShellOrientation(array_1d<double,3>& v1, //output
                             const array_1d<double,3>& v2, //input parameter: direction of 2 - 
                             const array_1d<double,3>& v3        //input parameter shell normal
                       )


     void Compute3DOrientation(   const array_1d<double,3>& v1, //input parameter: direction of 1
                             const array_1d<double,3>& v2, //input parameter: direction of 2  
                             array_1d<double,3>& v3        //output parameter v3 = v1 x v2
                       )



     //this function just computes the rotation matrix given an orthonormal basis 
     void ComputeRotationMatrix(  const array_1d<double,3>& v1, //input parameter: direction of 2
                             const array_1d<double,3>& v2, //input parameter: direction of 2  
                             const array_1d<double,3>& v3, //input parameter v3 = v1 x v2
                             bounded_matrix<double,3,3>& rot
                       )



     ##default case
    void ComputeDefaultBeamOrientation( array_1d<double,3>& v1, //output parameter
                             array_1d<double,3>& v2, //OUTPUT parameter: direction of 2
                             const array_1d<double,3>& v3 //input parameter direction of    the beam axis
                       )

    void ComputeDefaultShellOrientation(array_1d<double,3>& v1, //output
                             array_1d<double,3>& v2, //output parameter: direction of 2 - 
                             const array_1d<double,3>& v3  //output parameter v3 = v1 x v2
                       )


    void ComputeDefault3DOrientation(   array_1d<double,3>& v1, //output parameter: direction of 2
                             array_1d<double,3>& v2, //output parameter: direction of 2  
                             array_1d<double,3>& v3        //output parameter v3 = v1 x v2
                       )

than a base element for a BEAM would look something like:

    Initialize()
    {
       array_1d<double,3> v1;  //to be computed
       array_1d<double,3> v2; //to be computed
       array_1d<double,3> v3 = GetGeometry()[1].Coordinates() - GetGeometry()[0].Coordinates();
       if(this->Has(DIRECTION2))
       {
             v2 = this->GetValue(DIRECTION2);
             ComputeBeamOrientation(v1, v2, v3)
       }
       else //user did not define anything
       {
             ComputeDefaultBeamOrientation(v1,v2,v3); //here both v2 and v1 are computed
       }
       
       bounded_matrix<double,3,3> Rot;
       ComputeRotationMatrix(v1,v2,v3,Rot);
    }     

is at least this idea fine to everyone?

**Comment by msandre** [comment]
I think it's still consistent with orthonormalization for determining v2. One only needs to change my original post to define v1:= v2 x v3.

**Comment by RiccardoRossi** [comment]
@msandre ok, even better. than i think that there is agreement at least for the **BEAM** case.
that is

    v3 = normalize(x2 - x1)
    v2 to be either prescribed by the user or computed automatically
    v1 = v1 x v2

is it ok to everyone?
@josep-m-carbonell  is it ok to you?
@MassimoPetracca ?

**Comment by pooyan-dadvand** [comment]
I agree with the procedure in general but not with notation nor with implementation prototype.....

* @RiccardoRossi @msandre  I don't get your notation while it is the first time that I see the v1 is not in the direction of the beam. (please see the figure by Massimo where x' is in direction of the beam). IMO it is not standard and also confusing. 

* The interfaces are somehow rare. Usually the output is the last and here changes from one to other. (I would not pass the v3 as it is almost redundant)

* It is not clear what you want to store in the element. v1,.. or Rotation Matrix? if any of them is stored as a member I would not pass it to that functions.

* DIRECTION2 is not descriptive. I would use LOCAL_AXIS_Y instead

**Comment by RiccardoRossi** [comment]
ok implementation is to be discussed. Also where to put this functions is an open question.
my comments are:

- the choice of v3 instead of v1 for the beam axis is to have legacy code working. It is an attempt not to break the code of @josep-m-carbonell (and i think is also a common choice in some books). I am fine with having v1 in the direction of the beam axis, but it will break existing code.
- the function should **not** be a function of the element, since we must be able to use it also in the conditons and most probably in some other places. I think the correct placement is in an utility...
 - i would not use XYZ for the local axes since it identifies the global ones. instead i would prefer 123 or UVT. not a major issue to me though
     that is LOCAL_AXIS_U, LOCAL_AXED_V, LOCAL_AXES_T or
               LOCAL_AXIS_1, LOCAL_AXIS_2, LOCAL_AXIS_3

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi thanks for your comments!

In the same order:
* Now I see the point of v3. Thanks for explaining it. (I have missed it above) Still I don't like it but backward compatibility is also important. (Meanwhile, do we have any working beam in Kratos?)
* We will discuss the implementation details in another issue
* I agree with LOCAL_AXIS_1, ....

**Comment by josep-m-carbonell** [comment]
I don't know if I agree or not because you are mixing different things. The real problem is not the implementation.  It is the correspondence between the input and output local bases must have with the graphical interface.

**Comment by RiccardoRossi** [comment]
@josep-m-carbonell  what is the interface giving to you right now?
how is that information communicated to the element?

**Comment by josep-m-carbonell** [comment]
Right now you have the conectivities and  the properties,  which includes the section characteristics. That's all. 
Nothing about the local axes.

**Comment by josep-m-carbonell** [comment]
In fact, kratos elements are using v1 (LOCAL_AXIS_1) as beam axis, this is the used in the beam formulation.  The problem is that you must take into account that later on, GiD will expect a local axis where v3 is the beam axis instead of v1.

**Comment by josep-m-carbonell** [comment]
What I try to explain, is that the problem of keeping legacy is not exactly what happens here. Because if a new mechanism of supplying local axes (in a customized way)  is introduced,  the method to define the transformation matrix must be adapted to make the correct transfer. That it is always the same: what the interface defines in the model as local base, and what is considered as local base in the programed element.

**Comment by RiccardoRossi** [comment]
@josep-m-carbonell I honestly can not understand you :-(
let's imagine for a second we do not have constraints with backwards compatibility.
let's also assume that the user defines everything so that we have the 3 local axes identified.
we can write an element that complies with the new convection, and i guess we can do this correctly since we prescribed everything.

are u saying that the problem is at the moment of printing the results? i mean in the post processing...

if this is the case, can't we fix it by hacking the gid_io?

**Comment by josep-m-carbonell** [comment]
Yes, I can see that it is difficult to understand.
We are not talking about the local axes of a constraint.  We are talking about the local axes of a beam. (usually they coincide, of course, if you don't want to make things too complex). Therefore imagine that you have 100 elements. 
1) Are you going to prescribe all local axis for each element?
Usually not. You will use some default convention. 
In fact, the local axes will be defined over the geometry segments, not over the element. You would like to have the option to change some of these local axes. 
2) Well, now which is the input? The local axes for each beam element in the properties? You can have several properties, one for each beam, or at least for each geometry segment... Take into account that the local axes orthogonal to the beam axis will define the position of the beam section, which can be no symmetrical, IPN, HEB, L, T.... So it is important and must be controlled if it is possible. 
3) Internally the element can be programmed considering the more convenient directions for the programmer. Taking into account the supplied axes by the user (given by the interface preproces), because it defines the section orientation and the base were the internal forces must be expressed to return the results to the postprocess. 
4) Right now, we are not defining custom local axes, so we do not control the section orientation. It will be the one defined by the preprocess default convention. That for us is the GiD one, but could be another.

**Comment by RiccardoRossi** [comment]
Hi again,
        thank you for the explanation.

regarding your point 1, my plan is exactly to have a possibly different local axis **assigned to each beam element**, that is to assume that the vector LOCAL_AXIS_* is applied **to each element in the model**.

If the problemtype is not providing this information, than the first thing that the element should do is to calculated it internally, so that from that moment on it is exactly as if it was provided by the problemtype. This is the reason for which i was proposing the "*Default*" functions in my previous post.

an important advantageof this approach is that at any point one can ask to the beam to plot its local axis orientation, which is an important guide in the post process.


on a side note i just checked the documentation of GiD. you can find it by searching for the keyword "local_axes_direction" in the docs of GiD.

the important message is that they have an "automatic" mode which assigns 
LOCAL_AXES_1 --> in the direction of the beam axis
LOCAL_AXIS_2/3 --> orthogonal to it (the algorithm is described in the doc, but i think it more or less coincides with what was posted before

taking this into account i updated the code i posted before to take in account that the axis 1 is the beam axis

    Initialize()
    {
        const array_1d<double,3> v1;  GetGeometry()[1].Coordinates() - GetGeometry()[0].Coordinates();
        array_1d<double,3> v2; //to be computed
        array_1d<double,3> v3; //to be computed
        if(this->Has(LOCAL_AXIS_2))
        {
             v2 = this->GetValue(LOCAL_AXIS_2);
             ComputeBeamOrientation(v1, v2, v3)
        }
        else //user did not define anything - default is computed and saved
       {
            ComputeDefaultBeamOrientation(v1,v2,v3); //here both v2 and v3 are computed
            this->SetValue(LOCAL_AXIS_2, v2);
       }
   
       bounded_matrix<double,3,3> Rot;
       ComputeRotationMatrix(v1,v2,v3,Rot);
    } 

2) yes, this is exactly the reason for this thread
3) it is clear that the element internally could assume a different convention. However we can definitely decide a convention for IO
4) i addressed this in the point 1...

can we accept this? it does not break anything and it should result in the same IO convention as before

**Comment by msandre** [comment]
I would just make a few minor suggestions. Otherwise I agree.

1. I'm following nastran convention here. The user-provided beam orientation vector should be allowed to be non-orthogonal to beam axis. Initialize() should compute LOCAL_AXIS_2 from beam orientation vector and beam axis.

2. I prefer function names ComputeBeamLocalAxis2, ComputeBeamDefaultLocalAxis2 and ComputeBeamLocalAxis3 as they are more clear in my opinion.

3. I don't think this is a problem with the proposed interface as long as the ComputeBeam*() functions take v1, v2, v3 as arguments. But, we should be careful with assuming where LOCAL_AXIS_* are stored. It makes sense to me to associate these with a local point in the beam, say node A, which for a constant cross-section defines the orientation over the entire beam. Possibly there will be other beam where such properties are defined at multiple points along the length.

**Comment by msandre** [comment]
Or overloading...

void ComputeBeamLocalAxis2(const array_1d<double,3>& rLocalAxis1, array_1d<double,3>& rLocalAxis2)
void ComputeBeamLocalAxis2(const array_1d<double,3>& rLocalAxis1, const array_1d<double,3>& rOrientationVector, array_1d<double,3>& rLocalAxis2)
void ComputeBeamLocalAxis3(const array_1d<double,3>& rLocalAxis1, const array_1d<double,3>& rLocalAxis2, array_1d<double,3>& rLocalAxis3)

**Comment by RiccardoRossi** [comment]
@msandre,
      i am fine in principle with all of your remarks. Once we agree on the idea we'll start discussing the implementation details/naming etc

concerning your point 3, we could store it on integration point for maximal flexibility. (also @loumalouomega  was making me offline the same remark). What matters is that they should **not** be stored on nodes as they should be discontinuous.

**Comment by josep-m-carbonell** [comment]
I would like to make some corrections about my initial posts. I am reviewing the code programmed a couple of years ago, and I see that the orientation selected for the small displacement beam element is not the standard one.

The standard one can be found in the **beam_math_utilities.hpp** of the kratos core. Among them, there is a method called CalculateLocalAxesVectors(..) where the transformation is done.  It seams strange to me that GiD says that LocalAxis1 is the beam axis in his convention (as @RiccardoRossi posted before). If I type in GiD search "local axes definition" it says that beam axis is local-z in the CustomLib. In fact I took directly the code that is programmed into GiD to compute the three axes, and If I look into my comments everywhere says that beam Axis is LocalAxis3 in GiD. Threrefore I used this convention for the large displacement beam elements. 

So I can state that right now the standard is:
 - the beam Axis as LocalAxis3  (and the small displacement element is an exception and has a different transformation matrix)  

  - the inertia dyadic is something like

    //Axis Local E3
    rInertiaDyadic(0,0) = rSection.Inertia_y; //vertial axis
    rInertiaDyadic(1,1) = rSection.Inertia_z; //horizontal axis
    rInertiaDyadic(2,2) = rSection.Rotational_Inertia; //beam axis

(sorry for my initial confusion)

For the storage is better to use a quaternion for each integration point.

**Comment by RiccardoRossi** [comment]
![screenshot from 2017-03-13 18-56-11](https://cloud.githubusercontent.com/assets/8059665/23868199/d9ee2292-081e-11e7-807e-bbf351284fe8.png)

**Comment by RiccardoRossi** [comment]
are u proposing to use directly a quaternion for the initial axis configuration?

that could indeed be an option, however we must have functions (with a naming and format to be decided) that initialize it either by default OR by a prescribed axis if that is the case.

**Comment by josep-m-carbonell** [comment]
Yes I use a quaternion for the initial local axis transformation.  I think that the explanation in GiD is not correct....because I use the code they supplied me, that coincides with the CustomLib convention: 
![screenshot from 2017-03-13 19-17-29](https://cloud.githubusercontent.com/assets/23210788/23869025/e52c149a-0821-11e7-935f-f3120ca0c653.png)

We will have to ask them. Also the possibilities for the definition of the local axis and input entries.

**Comment by jginternational** [comment]
Kike Escolano (GiD main developer for preprocess) : GiD X' local axis goes from point 1 to point 2 in the line. It is not necessarily coinciding with the direction of the segment since it is computed as the local tangent to the underlying CAD geometry.
Of course there is a degree of indetermination in the "y" axis. A default convention is taken by GiD.

**Comment by josep-m-carbonell** [comment]
So this is something new, one year ago it was not like this. What about the customlib @jginternational, which are the default local axes ? Are they the same ?

**Comment by jginternational** [comment]
This GiD criteria has been like this since the origin of GiD, but I don't know how the former problemtype of kratos was writing this information.
I still need to study about customLib local axes. Anyway, the full customLib documentation about local axes can be found here:

http://www.gidhome.com/documents/customizationmanual/PROBLEMTYPE%20SYSTEM/Description%20of%20the%20local%20axes

**Comment by josep-m-carbonell** [comment]
Well, something does not match. I remember Enrique supplied me the code to coincide exactly with the one of GiD...is the postprocess following the same convention for writing the StressResultants and the StressCouples ?

It seems that CustomLib and GiD defaults does not match either. Doesn't it?

**Comment by RiccardoRossi** [comment]
@josep-m-carbonell 

in any case, and aside for the ongoing discussion on what is axis_1, i like the idea of using a quaternion to store the initial rotation matrix.

now let's imagine that my local basis is given by 
v1 = 1/sqrt(2), 1/sqrt(2), 0
v2 = -1/sqrt(2), 1/sqrt(2),0
v3 = 0 0 1

given this information how do i store it in a quaternion using your helper function?

**Comment by josep-m-carbonell** [comment]
You can get a Quaternion from a rotation vector or from a Rotation Matrix ( see **quaternion.h**). If you have the transformation matrix in columns LocalTransformationMatrix = [v1 | v2 | v3]
 `mInitialLocalQuaternion = QuaternionType::FromRotationMatrix(LocalTransformationMatrix);`

**Comment by RiccardoRossi** [comment]
ok, perfect, than i believe it is a clever solution.

@josep-m-carbonell or @MassimoPetracca could any of you add a short "How to use quaternions" to the github wiki?

later on we will unit test it (i can do that).

**Comment by josep-m-carbonell** [comment]
After talking with the GiD team we concluded that there is an error in the GiD documentation. We checked everything and the correct is: 

1) In preprocess they use LocalAxisX as beam axis.

2) There is a default calculation of the other axes (orthogonal axes), that is the one in the manual and the one reproduced in beam_math_utilities.hpp and in the SmallDisplacementBeamElement.

3) In postprocess they use global axes representation. So StressResultants and StressCouples must be returned to postprocess in global axes. There is no way to represent things in local axes right now.

We can use internally their convention or not.  Just we have to return internal forces in global axes.

**Comment by RiccardoRossi** [comment]
so, i added a very initial documentation of the conventions to the github wiki. 

please edit is as yo u consider.

@msandre @loumalouomega  i wrote in the documentation that LOCAL_AXES_2 is to be prescribed by SetValue on the element NOT on the gauss point. The rationale is that then you can compute it at the gauss points internally AND you need to be able to call "Has" to know if it was set or not

**Comment by RiccardoRossi** [comment]
what if we define a helper class "Orientation" (pick a better name as you like

and do

    class Orientation
    {
        public:
        Orientation(array_1d<double,3>& v1) //this constructor is to be used if only the direction of v1 is given. Default orientation will be assumed for the other axis
        {
          ...
        }

        Orientation(array_1d<double,3>& v1, //this constructor is to be used if the user has prescribed two axes
                array_1d<double,3>& v2
                )
        {
           ...
        }

        void GetRotationMatrix(bounded_matrix<double,3,3>& R){...}

        void GetBasisVectors(array_1d<double,3>& v1, array_1d<double,3>& v2, array_1d<double,3>& v3){...}

        quaternion& GetQuaternion(){return morientation;}

        private:
             
        quaternion morientation;

    }

**Comment by MassimoPetracca** [comment]
@RiccardoRossi I like the idea of having a helper class. It is the best way to follow a unified convention.
I would like to raise another point:
What happens with geometric nonlinearity ? The initial orientation changes. 
And its update should be done incrementally, i.e. you cannot use the "Orientation" class directly on the deformed configuration, otherwise you will have sudden changes in the "updated" orientation.
Do you agree?

**Comment by RiccardoRossi** [comment]
in the nonlinear case you could store two instances 

     initial_orientation
     current_orientation

if you need the relative rotation you then do 

     current_orientation.GetRotationMatix() * InvertMatrix( initial_orientation.GetRotationMatrix())

**Comment by RiccardoRossi** [comment]
@msandre could u put in the wiki the default choice for v2? you posted it here, but i would prefer to be sure to have it right

**Comment by MassimoPetracca** [comment]
@RiccardoRossi Yes this can be done of course (or even better 'transpose' instead of 'invert' since it is a proper orthogonal matrix, or even better quaternion.conjugate() which is the same but faster).
But what I mean is, how do you construct the current_orientation ? if you construct it
using the user-defined vector, that vector is inputed by the user in the undeformed configuration. It may happen that with very large rotation this vector will be (for example) aligned with the current beam axis...
I don't know, I'm just thinking about it...
Probably it would be better to construct the current_orientation by updating (somehow) the initial one at each time step...

**Comment by MassimoPetracca** [comment]
sorry I accidentally closed it...

**Comment by msandre** [comment]
@RiccardoRossi I added a definition of the default for v2. The choice of cos(10) is somewhat arbitrary.

**Comment by josep-m-carbonell** [comment]
@MassimoPetracca You have the rotation in nodes,  if you have the initial local axes  you can rotate them and you will have the beam current local axes in each node and you can interpolate to have them expressed in each integration point.

**Comment by oberbichler** [comment]
@MassimoPetracca I think you get the same problem when you evaluate a curved beam at multiple points. You will get a singularity if the tangent is parallel to the user-defined vector.

**Comment by AndreasWinterstein** [comment]
@RiccardoRossi We would like to implement the correct coordinate system now. Could you please summarize the outcome of the discussion, since I am currently not completely sure which of the systems suggested above should be chosen.
Thank you.

**Comment by RiccardoRossi** [comment]
Hi Andreas,

i documented the convention for the beam orientation in the wiki.

[Local-axes-orientation](https://github.com/KratosMultiphysics/Kratos/wiki/Local-axes-orientation
)
also i think that we have an agreement about the fact that quaternions should be used in storing the initial orientation.

@MassimoPetracca wrote a small documentation on how to use the quaternions in the github wiki

[quaternions](https://github.com/KratosMultiphysics/Kratos/wiki/How-to-use-quaternions-in-Kratos)

i think that we should implement a function like

      class Orientation
      {
        public:
        Orientation(array_1d<double,3>& v1) //this constructor is to be used if only the direction of v1 is     given. Default orientation will be assumed for the other axis
         {
           ...HERE WE SHOULD IMPLEMENT THIS "DEFAULT CASE" 
         }

      Orientation(array_1d<double,3>& v1, //this constructor is to be used if the user has prescribed two axes
            array_1d<double,3>& v2
            )
    {
       ...
    }

        void GetRotationMatrix(bounded_matrix<double,3,3>& R){...}

        void GetBasisVectors(array_1d<double,3>& v1, array_1d<double,3>& v2, array_1d<double,3>& v3){...}

        quaternion& GetQuaternion(){return morientation;}

        private:
         
            quaternion morientation;

      }

to make sure that everyone will ultimately use the same implementaiton.
Do you think your student could provide such implementation following @msandre's suggestions?

**Comment by pooyan-dadvand** [comment]
I like the idea of orientation class. Nevertheless the I would use Calculate instead of Get for RotationMatrix and Basis to announce that they are not simply returning an internal variable and they need to calculate it each time.

**Comment by KlausBSautter** [comment]
I agree with the use of Calculate instead of Get and implemented the class Orientation for my beam element. That was a good idea, thanks. Nevertheless I added another constuctor
Orientation(array_1d<double,3>& v1, const double theta)
as I think it's a good idea to have the default case with a custom angle representing a rotation around the local beam axis 1.

**Comment by RiccardoRossi** [comment]
Where do you define the angle from in this case?

That is, what is the orientación of v2 that corresponda to a given angle
theta?

**Comment by KlausBSautter** [comment]
` I wrote the following code for the case of a given theta. I have theta as a new KRATOS Variable that can be defined in the ProjectParameter.json file. Theta just rotates the default case around the local_axis_1

        Orientation::Orientation(array_1d<double, 3>& v1, const double theta) {
               KRATOS_TRY
		const int number_of_nodes = 2;
		const int dimension = 3;
		const int size = number_of_nodes * dimension;
		const int MatSize = 2 * size;

		array_1d<double, 3> GlobalZ = ZeroVector(dimension);
		GlobalZ[2] = 1.0;

		array_1d<double, 3> v2 = ZeroVector(dimension);
		array_1d<double, 3> v3 = ZeroVector(dimension);

		double VectorNorm;
		VectorNorm = MathUtils<double>::Norm(v1);
		if (VectorNorm != 0) v1 /= VectorNorm;

		if (fabs(v1[2]) > cos(10 * 2 * PI / 360)) {

			if (v1[2] > 0.00) {
				v1 = ZeroVector(dimension);
				v1[2] = 1.0;
				v2[1] = 1.0;
				v3[0] = -1.0;
			}

			else if (v1[2] < 0.00) {
				v1 = ZeroVector(dimension);
				v1[2] = -1.0;
				v2[1] = 1.0;
				v3[0] = 1.0;
			}
		}
		else {
			v2 = MathUtils<double>::CrossProduct(v1, GlobalZ);
			VectorNorm = MathUtils<double>::Norm(v2);
			v2 /= VectorNorm;

			v3 = MathUtils<double>::CrossProduct(v2, v1);
			VectorNorm = MathUtils<double>::Norm(v3);
			if (VectorNorm != 0) v3 /= VectorNorm;
		}

		//manual rotation around the beam axis
		const Vector nz_temp = v3;
		const Vector ny_temp = v2;
		if (theta != 0) {
			v2 = ny_temp * cos(theta) + nz_temp * sin(theta);
			v2 /= MathUtils<double>::Norm(v2);

			v3 = nz_temp * cos(theta) - ny_temp * sin(theta);
			v3 /= MathUtils<double>::Norm(v3);
		}

		Matrix RotationMatrix = ZeroMatrix(dimension);
		for (int i = 0; i < dimension; ++i) {
			RotationMatrix(i, 0) = v1[i];
			RotationMatrix(i, 1) = v2[i];
			RotationMatrix(i, 2) = v3[i];
		}


		this->GetQuaternion() = Quaternion<double>::FromRotationMatrix(RotationMatrix);

		KRATOS_CATCH("")
	}`

**Comment by RiccardoRossi** [comment]
Hi Klaus,
 i would suggest to define the function as 

      Orientation::Orientation(array_1d<double, 3>& v1, const double theta = 0.0)

that is to default the theta to 0.0. This way the parameters is optional and the implementation is only one.

On the optimization side, you could avoid computing multiple times the cos and sin of theta since it is a quite expensive operation. Same thing for computing the norm of v3 only once.

i would also ask you to explain a bit more extensively the default orientation in the wiki, so that it remains documented and one does not need to check the code. You can find the link here:

https://github.com/KratosMultiphysics/Kratos/wiki/Local-axes-orientation

btw, the rule coincides finally with the one of GiD? If this is not the case the documentation should ideally explain the differences

**Comment by KlausBSautter** [comment]
Hi, I already assigned a default value to theta as you propose in the decleration of the constructor. For the same reasons you mentioned.
Thanks for the hint I'll change the cos and sin calculations and talk to Andreas about the documentation when I see him tmr.

**Comment by RiccardoRossi** [comment]
very minor:

add a const to array_1d<double, 3>& v1

**Comment by KlausBSautter** [comment]
As I am doing this 


     if (fabs(v1[2]) > cos(10 * 2 * PI / 360)) {

		if (v1[2] > 0.00) {
			v1 = ZeroVector(dimension);
			v1[2] = 1.0;
			v2[1] = 1.0;
			v3[0] = -1.0;
		}

		else if (v1[2] < 0.00) {
			v1 = ZeroVector(dimension);
			v1[2] = -1.0;
			v2[1] = 1.0;
			v3[0] = 1.0;
		}

I dont think a const would work here + I am normalising v1 as one of the first steps

**Comment by RiccardoRossi** [comment]
ok, you are right

**Comment by RiccardoRossi** [comment]
Excuse me again, but i now read your code and i am a little perplex.

v1 := (x2 - x1) / norm(x2 - x1)

**you can never change its direction!**

only thing that should be changed in the ifs is the direction of v2 and consequently of v3, but you can not modify the orientation of the beam axis 1

**Comment by KlausBSautter** [comment]
yes my idea was to say that if e.g. v1 = [0.1,0.1,0.98] we say that this is nearly [0,0,1] to make things easier . this is why I write
'if (fabs(v1[2]) > cos(10 * 2 * PI / 360)) {'
But I can just get rid of this if you dont like that.

update: this would be a too big simplification especially for small incremental deformation..., I will change it
